### ISSUE TITLE
TaskSupport overridden in ForkJoinTasks

### ISSUE THREAD

**Issue by scabug** [scala/scala-parallel-collections#160]
The wrong forkJoinPool is used in the following example:

```scala
implicit val someForkJoinPoolExecutionContext = ...
val otherForkJoinPoolEC = ...
val someCollection = ...
def foo() {
   // Future runs in someForkJoinPoolExecutionContext
   Future{
       val parCollection = someCollection.par
       parCollection.tasksupport = new ExecutionContextTaskSupport(otherForkJoinPoolEC)
       // this is executed in the someForkJoinPoolExecutionContext rather than otherForkJoinPoolEC
       parCollection.foreach(...)
    }
} 
```

In Tasks.scala, we see at line 420: 
https://github.com/scala/scala/blob/5cb3d4ec14488ce2fc5a1cc8ebdd12845859c57d/src/library/scala/collection/parallel/Tasks.scala#L420

```scala
if (Thread.currentThread.isInstanceOf[ForkJoinWorkerThread]) {
```

Which overrides the given tasksupport, preferring the executioncontext of the currentThread.

**Comment by michellemay** [comment]
> The umbrella issue for TaskSupport ergonomics might be #152
> 
> Probably @SethTisue intended to close this ticket with other module-related tickets?
> 
> I tried to follow the conclusions from a few years ago. For `par`, using the `ForkJoinPool` ctor with `maximumPoolSize` and `minimumRunnable` isn't sufficient to limit the "outer loop" to 2 submitted tasks and max 2 threads.
> 
> A workaround is to call `sum` inside `Future` (using default context aka common pool). (When the inner par operation is run, on the common pool I assume, is that run on the same thread? is it a fork or submit?)
> 
> Here is the updated snippet. The printlns make it easier to see order of operations. In the absence of blocking, the outer loop happily submits tasks, and the FJP may spawn an extra thread or 2 to ensure parallelism. It's crucial to distinguish tasks from the threads. I assume the expectation or goal is to run 2 tasks on 2 threads, with the next task (or foreach iteration) running when a thread is available.
> 

This is an interesting solution but not one that is possible given the current code I use.  The inner loop is not directly accessible. 
 It would require to change all method signatures and objects to add `commonsupport` parameter.  It does not make sense since the inner code might run parallel operations anywhere.  

My workaround  is kind of the opposite of your code sample..  I execute the top level loop with futures on a fixed execution pool and leave the inner code do whatever it wants.

**Comment by som-snytt** [comment]
I agree your solution is best. The promise of `.par` was that I can sprinkle it around and benefit.
